File: src/mesh/LightGrid.js
/**
 * LightGrid - Sparse storage for per-block light values
 * 
 * Stores sky light (0-15) for each block position using the same
 * section-based sparse storage pattern as BinaryGrid.
 * 
 * Minecraft's sky light at noon:
 * - 15 = Full sunlight (outdoors)
 * - 0 = Complete darkness (deep caves)
 * - Decrements by 1 per block in shadow
 */

import { 
  SECTION_SIZE, 
  SECTION_VOLUME,
  MIN_Y,
  makeSectionKey,
  parseSectionKey,
  worldYToSection,
  blockIndexInSection
} from './BinaryGrid.js';

// Maximum light level (full sunlight at noon)
export const MAX_LIGHT = 15;

/**
 * LightGrid class - sparse storage for light values
 */
export class LightGrid {
  constructor() {
    // Map of section key -> Uint8Array(4096)
    // Each byte stores sky light (0-15) in lower nibble
    // Upper nibble reserved for block light if needed later
    this.sections = new Map();
    
    // Track bounds for efficient iteration
    this.minChunkX = Infinity;
    this.maxChunkX = -Infinity;
    this.minChunkZ = Infinity;
    this.maxChunkZ = -Infinity;
    this.minSectionY = Infinity;
    this.maxSectionY = -Infinity;
    
    // Flag to track if this grid has pre-loaded Minecraft light data
    // If true, missing sections should default to 0 (dark underground)
    // If false, we're using fallback propagation or empty data
    this.hasMinecraftLightData = false;
  }
  
  /**
   * Clear all data
   */
  clear() {
    this.sections.clear();
    this.minChunkX = Infinity;
    this.maxChunkX = -Infinity;
    this.minChunkZ = Infinity;
    this.maxChunkZ = -Infinity;
    this.minSectionY = Infinity;
    this.maxSectionY = -Infinity;
    this.hasMinecraftLightData = false;
  }
  
  /**
   * Get or create a section
   */
  _getOrCreateSection(chunkX, chunkZ, sectionY) {
    const key = makeSectionKey(chunkX, chunkZ, sectionY);
    let section = this.sections.get(key);
    
    if (!section) {
      section = new Uint8Array(SECTION_VOLUME);
      this.sections.set(key, section);
      
      // Update bounds
      if (chunkX < this.minChunkX) this.minChunkX = chunkX;
      if (chunkX > this.maxChunkX) this.maxChunkX = chunkX;
      if (chunkZ < this.minChunkZ) this.minChunkZ = chunkZ;
      if (chunkZ > this.maxChunkZ) this.maxChunkZ = chunkZ;
      if (sectionY < this.minSectionY) this.minSectionY = sectionY;
      if (sectionY > this.maxSectionY) this.maxSectionY = sectionY;
    }
    
    return section;
  }
  
  /**
   * Get a section (returns undefined if not exists)
   */
  getSection(chunkX, chunkZ, sectionY) {
    const key = makeSectionKey(chunkX, chunkZ, sectionY);
    return this.sections.get(key);
  }
  
  /**
   * Set sky light at world coordinates
   * @param {number} worldX - World X coordinate
   * @param {number} worldY - World Y coordinate
   * @param {number} worldZ - World Z coordinate
   * @param {number} skyLight - Sky light level (0-15)
   */
  setSkyLight(worldX, worldY, worldZ, skyLight) {
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    const section = this._getOrCreateSection(chunkX, chunkZ, sectionY);
    const index = blockIndexInSection(localX, localY, localZ);
    
    // Store sky light in lower nibble (preserving upper nibble for block light)
    section[index] = (section[index] & 0xF0) | (skyLight & 0x0F);
  }
  
  /**
   * Get sky light at world coordinates
   * @param {number} worldX - World X coordinate
   * @param {number} worldY - World Y coordinate
   * @param {number} worldZ - World Z coordinate
   * @returns {number} Sky light level (0-15)
   */
  getSkyLight(worldX, worldY, worldZ) {
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    const section = this.getSection(chunkX, chunkZ, sectionY);
    // If section doesn't exist, use same logic as getLight()
    if (!section) {
      const withinBounds = 
        chunkX >= this.minChunkX && chunkX <= this.maxChunkX &&
        chunkZ >= this.minChunkZ && chunkZ <= this.maxChunkZ;
      
      if (this.hasMinecraftLightData && withinBounds) {
        return 0; // Underground with MC data
      }
      return MAX_LIGHT; // Outside bounds or no MC data
    }
    
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    const index = blockIndexInSection(localX, localY, localZ);
    return section[index] & 0x0F;
  }
  
  /**
   * Set block light at world coordinates
   * @param {number} worldX - World X coordinate
   * @param {number} worldY - World Y coordinate
   * @param {number} worldZ - World Z coordinate
   * @param {number} blockLight - Block light level (0-15)
   */
  setBlockLight(worldX, worldY, worldZ, blockLight) {
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    const section = this._getOrCreateSection(chunkX, chunkZ, sectionY);
    const index = blockIndexInSection(localX, localY, localZ);
    
    // Store block light in upper nibble (preserving lower nibble for sky light)
    section[index] = (section[index] & 0x0F) | ((blockLight & 0x0F) << 4);
  }
  
  /**
   * Get block light at world coordinates
   * @param {number} worldX - World X coordinate
   * @param {number} worldY - World Y coordinate
   * @param {number} worldZ - World Z coordinate
   * @returns {number} Block light level (0-15), defaults to 0 if not set
   */
  getBlockLight(worldX, worldY, worldZ) {
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    const section = this.getSection(chunkX, chunkZ, sectionY);
    if (!section) return 0;
    
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    const index = blockIndexInSection(localX, localY, localZ);
    return (section[index] >> 4) & 0x0F;
  }
  
  /**
   * Get both sky light and block light at world coordinates
   * @param {number} worldX - World X coordinate
   * @param {number} worldY - World Y coordinate
   * @param {number} worldZ - World Z coordinate
   * @returns {{skyLight: number, blockLight: number}} Light levels (0-15 each)
   */
  getLight(worldX, worldY, worldZ) {
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    const section = this.getSection(chunkX, chunkZ, sectionY);
    // If section doesn't exist, we need to determine if this is:
    // 1. Empty air above terrain (should be bright, sky=15)
    // 2. Underground area with missing data (should be dark, sky=0)
    // 3. Outside loaded chunk bounds (default to sky=15 for safety)
    if (!section) {
      // Check if this chunk position is within our loaded bounds
      // If we have Minecraft light data and the chunk is within bounds,
      // a missing section likely means underground (dark)
      // If outside bounds, default to bright to avoid darkening unloaded areas
      const withinBounds = 
        chunkX >= this.minChunkX && chunkX <= this.maxChunkX &&
        chunkZ >= this.minChunkZ && chunkZ <= this.maxChunkZ;
      
      if (this.hasMinecraftLightData && withinBounds) {
        // Within loaded chunks with MC data - missing section = underground
        return { skyLight: 0, blockLight: 0 };
      }
      // Outside bounds or no MC data - default to bright (safe fallback)
      return { skyLight: MAX_LIGHT, blockLight: 0 };
    }
    
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    const index = blockIndexInSection(localX, localY, localZ);
    const value = section[index];
    return {
      skyLight: value & 0x0F,
      blockLight: (value >> 4) & 0x0F,
    };
  }
  
  /**
   * Fast sky light lookup using pre-computed section
   * @param {Uint8Array} section - The light section
   * @param {number} localX - Local X (0-15)
   * @param {number} localY - Local Y (0-15)
   * @param {number} localZ - Local Z (0-15)
   * @returns {number} Sky light level (0-15)
   */
  getSkyLightInSection(section, localX, localY, localZ) {
    const index = localY * SECTION_SIZE * SECTION_SIZE + localZ * SECTION_SIZE + localX;
    return section[index] & 0x0F;
  }
  
  /**
   * Set sky light using pre-computed section (for batch operations)
   */
  setSkyLightInSection(section, localX, localY, localZ, skyLight) {
    const index = localY * SECTION_SIZE * SECTION_SIZE + localZ * SECTION_SIZE + localX;
    section[index] = (section[index] & 0xF0) | (skyLight & 0x0F);
  }
  
  /**
   * ULTRA-FAST: Get light value by direct index (no coordinate conversion)
   * @param {Uint8Array} section - The light section
   * @param {number} index - Direct array index (0-4095)
   * @returns {number} Combined light byte (sky in lower nibble, block in upper)
   */
  static getLightByIndex(section, index) {
    return section[index];
  }
  
  /**
   * ULTRA-FAST: Get sky and block light by index (inline-friendly)
   * Call as: const val = section[index]; skyLight = val & 0xF; blockLight = val >> 4;
   */
  
  /**
   * Get a light section by section key directly (avoids coordinate math)
   * @param {string} key - Section key from makeSectionKey
   * @returns {Uint8Array|undefined}
   */
  getSectionByKey(key) {
    return this.sections.get(key);
  }
  
  /**
   * Get memory usage estimate
   */
  getMemoryUsage() {
    return {
      sectionCount: this.sections.size,
      bytesUsed: this.sections.size * SECTION_VOLUME,
    };
  }

  /**
   * Export grid data for worker transfer
   * Returns data that can be sent to workers via postMessage
   */
  export() {
    const sectionsData = [];
    
    for (const [key, section] of this.sections) {
      sectionsData.push({
        key,
        data: section.buffer.slice(0), // Copy buffer
      });
    }
    
    return {
      sections: sectionsData,
      bounds: {
        minChunkX: this.minChunkX,
        maxChunkX: this.maxChunkX,
        minChunkZ: this.minChunkZ,
        maxChunkZ: this.maxChunkZ,
        minSectionY: this.minSectionY,
        maxSectionY: this.maxSectionY,
      },
    };
  }

  /**
   * Import grid data from export
   */
  static import(data) {
    const grid = new LightGrid();
    
    for (const { key, data: buffer } of data.sections) {
      const section = new Uint8Array(buffer);
      grid.sections.set(key, section);
    }
    
    if (data.bounds) {
      grid.minChunkX = data.bounds.minChunkX;
      grid.maxChunkX = data.bounds.maxChunkX;
      grid.minChunkZ = data.bounds.minChunkZ;
      grid.maxChunkZ = data.bounds.maxChunkZ;
      grid.minSectionY = data.bounds.minSectionY;
      grid.maxSectionY = data.bounds.maxSectionY;
    }
    
    return grid;
  }
}

export default LightGrid;

File: src/wasm-mesher/src/grid/light_grid.rs
//! LightGrid - Sparse storage for per-block light values
//!
//! Each byte stores:
//! - Lower 4 bits: Sky light (0-15)
//! - Upper 4 bits: Block light (0-15)

use std::collections::HashMap;
use crate::types::{
    SectionKey, SECTION_SIZE, SECTION_VOLUME, block_index_in_section, world_y_to_section,
};

/// Maximum light level
pub const MAX_LIGHT: u8 = 15;

/// Light values for a single block
#[derive(Debug, Clone, Copy, Default)]
pub struct LightValue {
    pub sky_light: u8,
    pub block_light: u8,
}

impl LightValue {
    pub fn new(sky: u8, block: u8) -> Self {
        Self {
            sky_light: sky.min(MAX_LIGHT),
            block_light: block.min(MAX_LIGHT),
        }
    }

    /// Pack into single byte
    pub fn pack(&self) -> u8 {
        (self.block_light << 4) | self.sky_light
    }

    /// Unpack from single byte
    pub fn unpack(byte: u8) -> Self {
        Self {
            sky_light: byte & 0x0F,
            block_light: (byte >> 4) & 0x0F,
        }
    }
}

/// Sparse light grid
pub struct LightGrid {
    /// Sections stored by packed key
    sections: HashMap<u64, Box<[u8; SECTION_VOLUME]>>,
    /// True if this grid contains Minecraft's pre-computed light data
    /// When true, missing sections default to 0 (dark)
    /// When false (fallback mode), missing sections default to MAX_LIGHT
    has_minecraft_data: bool,
    /// Public flag for decode module to set
    pub has_minecraft_light_data: bool,
}

impl LightGrid {
    pub fn new() -> Self {
        Self {
            sections: HashMap::new(),
            has_minecraft_data: false,
            has_minecraft_light_data: false,
        }
    }

    /// Import from serialized byte data
    /// Format: [num_sections: u32][section_key: u64, data: [u8; 4096]]...
    pub fn from_bytes(data: &[u8]) -> Self {
        let mut grid = Self::new();
        
        if data.len() < 4 {
            return grid;
        }

        let num_sections = u32::from_le_bytes([data[0], data[1], data[2], data[3]]) as usize;
        let mut offset = 4;

        for _ in 0..num_sections {
            if offset + 8 + SECTION_VOLUME > data.len() {
                break;
            }

            // Read section key (packed u64)
            let key = u64::from_le_bytes([
                data[offset], data[offset + 1], data[offset + 2], data[offset + 3],
                data[offset + 4], data[offset + 5], data[offset + 6], data[offset + 7],
            ]);
            offset += 8;

            // Read section data
            let mut section = Box::new([0u8; SECTION_VOLUME]);
            section.copy_from_slice(&data[offset..offset + SECTION_VOLUME]);
            offset += SECTION_VOLUME;

            grid.sections.insert(key, section);
        }

        // If we loaded any sections, we have Minecraft light data
        // Missing sections should default to 0 (dark), not MAX_LIGHT
        grid.has_minecraft_data = !grid.sections.is_empty();

        grid
    }

    /// Get a section by key
    #[inline]
    pub fn get_section(&self, key: &SectionKey) -> Option<&[u8; SECTION_VOLUME]> {
        self.sections.get(&key.to_packed()).map(|s| s.as_ref())
    }

    /// Get a section by packed key
    #[inline]
    pub fn get_section_by_packed(&self, packed: u64) -> Option<&[u8; SECTION_VOLUME]> {
        self.sections.get(&packed).map(|s| s.as_ref())
    }

    /// Get light at world coordinates
    #[inline]
    pub fn get_light(&self, world_x: i32, world_y: i32, world_z: i32) -> LightValue {
        let chunk_x = world_x.div_euclid(SECTION_SIZE as i32);
        let chunk_z = world_z.div_euclid(SECTION_SIZE as i32);
        let section_y = world_y_to_section(world_y);

        let local_x = world_x.rem_euclid(SECTION_SIZE as i32) as usize;
        let local_y = world_y.rem_euclid(SECTION_SIZE as i32) as usize;
        let local_z = world_z.rem_euclid(SECTION_SIZE as i32) as usize;

        let key = SectionKey::new(chunk_x, chunk_z, section_y);
        if let Some(section) = self.get_section(&key) {
            let idx = block_index_in_section(local_x, local_y, local_z);
            LightValue::unpack(section[idx])
        } else {
            // If we have Minecraft light data, missing sections are dark (caves, unlit areas)
            // If we don't have data (fallback mode), assume full sky light (outdoor)
            if self.has_minecraft_data {
                LightValue::new(0, 0)
            } else {
                LightValue::new(MAX_LIGHT, 0)
            }
        }
    }

    /// Get sky light at world coordinates
    #[inline]
    pub fn get_sky_light(&self, world_x: i32, world_y: i32, world_z: i32) -> u8 {
        self.get_light(world_x, world_y, world_z).sky_light
    }

    /// Get block light at world coordinates
    #[inline]
    pub fn get_block_light(&self, world_x: i32, world_y: i32, world_z: i32) -> u8 {
        self.get_light(world_x, world_y, world_z).block_light
    }

    /// Iterate over all sections
    pub fn iter_sections(&self) -> impl Iterator<Item = (SectionKey, &[u8; SECTION_VOLUME])> {
        self.sections.iter().map(|(packed, section)| {
            (SectionKey::from_packed(*packed), section.as_ref())
        })
    }

    /// Get neighbor section
    #[inline]
    pub fn get_neighbor_section(
        &self,
        key: &SectionKey,
        dx: i32,
        dy: i32,
        dz: i32,
    ) -> Option<&[u8; SECTION_VOLUME]> {
        let neighbor_key = SectionKey::new(
            key.chunk_x + dx,
            key.chunk_z + dz,
            key.section_y + dy,
        );
        self.get_section(&neighbor_key)
    }

    /// Check if grid has any sections
    pub fn is_empty(&self) -> bool {
        self.sections.is_empty()
    }

    /// Get or create a section for writing
    /// Returns a mutable reference to the section data
    pub fn get_or_create_section(&mut self, chunk_x: i32, chunk_z: i32, section_y: i32) -> &mut [u8; SECTION_VOLUME] {
        let key = SectionKey::new(chunk_x, chunk_z, section_y);
        let packed = key.to_packed();
        
        self.sections.entry(packed)
            .or_insert_with(|| Box::new([0u8; SECTION_VOLUME]))
            .as_mut()
    }
}

impl Default for LightGrid {
    fn default() -> Self {
        Self::new()
    }
}

